tsdb/store.go

```diff
@@ -1278,6 +1278,79 @@ func (s *Store) ShardRelativePath(id uint64) (string, error) {
 	return relativePath(s.path, shard.path)
 }
 
+// DeleteSeries loops through the local shards and deletes the series data for
+// the passed in series keys.
+func (s *Store) DeleteSeriesWithPredicate(database string, min, max int64, pred influxdb.Predicate) error {
+	s.mu.RLock()
+	if s.databases[database].hasMultipleIndexTypes() {
+		s.mu.RUnlock()
+		return ErrMultipleIndexTypes
+	}
+	sfile := s.sfiles[database]
+	if sfile == nil {
+		s.mu.RUnlock()
+		// No series file means nothing has been written to this DB and thus nothing to delete.
+		return nil
+	}
+	shards := s.filterShards(byDatabase(database))
+	epochs := s.epochsForShards(shards)
+	s.mu.RUnlock()
+
+	// Limit to 1 delete for each shard since expanding the measurement into the list
+	// of series keys can be very memory intensive if run concurrently.
+	limit := limiter.NewFixed(1)
+
+	return s.walkShards(shards, func(sh *Shard) error {
+		limit.Take()
+		defer limit.Release()
+
+		// install our guard and wait for any prior deletes to finish. the
+		// guard ensures future deletes that could conflict wait for us.
+		waiter := epochs[sh.id].WaitDelete(newGuard(min, max, nil, nil))
+		waiter.Wait()
+		defer waiter.Done()
+
+		index, err := sh.Index()
+		if err != nil {
+			return err
+		}
+		// indexSet := IndexSet{Indexes: []Index{index}, SeriesFile: sfile}
+
+		// Find matching series keys for each measurement.
+		mitr, err := index.MeasurementIterator()
+		if err != nil {
+			return err
+		}
+		defer mitr.Close()
+
+		for {
+			mm, err := mitr.Next()
+			if err != nil {
+				return err
+			} else if mm == nil {
+				break
+			}
+
+			if err := func() error {
```

Comment:
Readability question: is this nested function needed? Could we improve readability by inlining the body? I see we'd need to:
1. Change he `return nil` on line 1339 to a `break`
2. Refactor the call to `sh.DeleteSeriesRange` to be `if err := sh.DeleteSeriesRange(...); err != nil { return err }`

Reply:
The closure is there to ensure the `defer sitr.Close()` executes since it's in a `for` loop. Otherwise all the defers would stack up at the end of the function passed to `walkShards()`.

Reply:
Aha, makes sense to me. Thanks!